src/solana/solanaEngine.js

```diff
@@ -0,0 +1,390 @@
+
+const ACCOUNT_POLL_MILLISECONDS = 5000
+const BLOCKCHAIN_POLL_MILLISECONDS = 20000
+const TRANSACTION_POLL_MILLISECONDS = 3000
+
+export class SolanaEngine extends CurrencyEngine {
+  keypair: Keypair
+  feePerSignature: string
+  recentBlockhash: string
+  chainCode: string
+  otherData: SolanaOtherData
+  fetchCors: EdgeFetchFunction
+  settings: SolanaSettings
+
+  constructor(
+    currencyPlugin: SolanaPlugin,
+    walletInfo: EdgeWalletInfo,
+    opts: any, // EdgeCurrencyEngineOptions
+    fetchCors: EdgeFetchFunction
+  ) {
+    super(currencyPlugin, walletInfo, opts)
+    this.keypair = currencyPlugin.createKeyPair(
+      this.walletInfo.keys[`${this.currencyPlugin.pluginId}Mnemonic`]
+    )
+    this.chainCode = currencyPlugin.currencyInfo.currencyCode
+    this.fetchCors = fetchCors
+    this.feePerSignature = '5000'
+    this.recentBlockhash = '' // must be < ~2min old to send tx
+    this.settings = currencyPlugin.currencyInfo.defaultSettings.otherSettings
+  }
+
+  async fetchRpc(method: string, params: any = []) {
+    const body = {
+      jsonrpc: '2.0',
+      id: 1,
+      method,
+      params
+    }
+    const options = {
+      method: 'POST',
+      headers: {
+        Accept: 'application/json',
+        'Content-Type': 'application/json'
+      },
+      body: JSON.stringify(body)
+    }
+
+    const funcs = this.settings.rpcNodes.map(serverUrl => async () => {
+      const res = await this.fetchCors(serverUrl, options)
+      if (!res.ok) {
+        throw new Error(
+          `fetchRpc ${options.method} failed error: ${res.status}`
+        )
+      }
+      return res.json()
+    })
+
+    const response = await asyncWaterfall(funcs)
+    return response.result
+  }
+
+  updateBalance(tk: string, balance: string) {
+    if (typeof this.walletLocalData.totalBalances[tk] === 'undefined') {
+      this.walletLocalData.totalBalances[tk] = '0'
+    }
+    if (!bns.eq(balance, this.walletLocalData.totalBalances[tk])) {
+      this.walletLocalData.totalBalances[tk] = balance
+      this.warn(`${tk}: token Address balance: ${balance}`)
+      this.currencyEngineCallbacks.onBalanceChanged(tk, balance)
+    }
+    this.tokenCheckBalanceStatus[tk] = 1
+    this.updateOnAddressesChecked()
+  }
+
+  async queryBalance() {
+    try {
+      const response = await this.fetchRpc('getBalance', [
+        this.keypair.publicKey.toBase58(),
+        { commitment: this.settings.commitment }
+      ])
+      const balance = asRpcBalance(response)
+      this.updateBalance(this.chainCode, balance.value.toString())
+    } catch (e) {
+      // fetching of account balances for uninitiated accounts throws error
+      if (
+        this.tokenCheckTransactionsStatus[this.chainCode] === 1 &&
+        this.transactionList[this.chainCode].length === 0
+      ) {
+        this.updateBalance(this.chainCode, '0')
+      }
+      this.error(`Error checking ${this.chainCode} address balance`, e)
```

Comment:
This isn't really an error if the account is uninitiated. Can you differentiate a true error and only log that?